multiqc/modules/whatshap/whatshap.py

```diff
@@ -0,0 +1,321 @@
+                - Replace 'nan' values with zero
+            """
+            for key, value in data.items():
+                # Try to make the value an int
+                try:
+                    data[key] = int(data[key])
+                except ValueError:
+                    pass
+
+                # Replace 'nan' with zero
+                if value == "nan":
+                    data[key] = 0
+
+        file_content = logfile["f"]
+        # This will later be replaced by the sample name from the file, unless
+        # we are parsing an empty file
+        sample = self.clean_s_name(logfile["s_name"], logfile["root"])
+        # Get the header and remove the "#" character from the first field
+        header = next(file_content).strip().split()
+        header[0] = header[0][1:]
+
+        # The results from this log file, a dictionary for every chromosome
+        results = defaultdict(dict)
+        # Parse the lines that have the data
+        for line in file_content:
+            spline = line.strip().split()
+            data = {field: value for field, value in zip(header, spline)}
+            process_data(data)
+            # Remove the sample and chromsome from the data
+            sample = data.pop("sample")
+            chromosome = data.pop("chromosome")
+            # Insert the current line under chromosome
+            results[chromosome] = data
+
+        # If we were parsing an empty file, we return dummy data (basically all
+        # zero) so the sample statistics are consistent, which eases parsing
+        # later on. WhatsHap uses the fictional 'ALL' chromsome to store the
+        # summary for all results, so we re-use that to store all zero's
+        if not results:
+            results = dict()
+            results["ALL"] = defaultdict(int)
+
+        return sample, results
+
+    @staticmethod
+    def get_summary_field(sample_stats):
+        """
+        Return the summary field for sample_stats
+
+        If there is only a single chromosome, we use that as the summary field.
+        If there are multiple chromosomes, whatshap adds the 'ALL' field which
+        contains the summary
+        """
+        # If there is only a single chromosome we use that as the summary
+        # filed. Otherwise, we use the 'ALL' chromosome (it only gets added
+        # by WhatsHap stats when there are multiple chromosomes).
+        if len(sample_stats) == 1:
+            summary_field = list(sample_stats)[0]
+        else:
+            summary_field = "ALL"
+        return summary_field
+
+    def whatshap_add_general_stats(self):
+        """Add WhatsHap stats to the general statistics table"""
+
+        # Store the configuration for each column, this is also used to
+        # determine which columns to add to the general statistics table
+        # https://whatshap.readthedocs.io/en/latest/guide.html#the-tsv-statistics-format
+        general_stats_headers = OrderedDict(
```

Comment:
This is a lot of columns for General Statistics, even if many are hidden by default. Is it worth moving most of these to a dedicated table (where they need not be hidden by default), leaving just one or two here which would be useful to compare to results from other tools?

Reply:
I'm not sure which columns would be most useful for users. I put them in the general statistics table in case people want to easily compare the results with other tools, for example to relate the number of phased variants to the % of duplicates from picard markduplicates.

Reply:
@marcelm Do you have any suggestions on which fields would be most useful for users to have in the General Statistics table? In our [PacBio pipeline](https://github.com/LUMC/PacBio-variantcalling), we are using `Input Variants`, `Unphased Variants`, `Phased Variants`, `Total Phased bp` and `Avg bp per Block`. Even though they are all correlated to how "good" the phasing is.

Reply:
(Disclaimer: I don’t actually use `whatshap stats` so much.)

If you were to force me to pick only two columns, I would probably choose
- the percentage of phased variants, something like `phased / heterozygous_variants`
- and something describing the block sizes in terms of bp such as median bp per block (`bp_per_block_median`) or the block NG50.

Reply:
What about the following three columns for the general statistics table?
1. `Input Variants`, which at least for my data differs wildly between samples, so that is useful context to have.
2. `Phased Variants`
3. `bp_per_block_avg`. 
I'll also add an additional table to the WhatsHap section of the MultiQC report that holds all statistics. I'll hide the stats that are present in the general statistics table by default to keep the table clean

Reply:
Not sure if the question was directed at me. I think it is your call as you have more experience looking at these statistics.

My thought was that it would be relevant to see how many of the variants that can in principle be phased (heterozygous variants) were actually phased. If there’s only `Phased variants` and `Input variants`, then I cannot say this because `Input variants` includes the homozygous variants. And my reason for suggesting a percentage was that I tend to estimate the ratio in my head anyway when I see two absolute numbers of which one describes a subset of the other.
On the other hand, if the coverage is sufficiently high to cover most of the heterozygous variants, WhatsHap will phase almost all of them and the ratio `phased / heterozygous_variants` is probably close to 100% most of the time and won’t be helpful.

One could also argue that `whatshap stats` columns should only show information relevant to phasing and therefore the variant count should exclude the homozygous variants. But it is of course convenient to see the full variant count.

`bp_per_block_avg` versus `_median` is probably a matter of taste.

As I said, I’m not the one who has to look at this in practice, so I’m fine with whichever way you think is best.

Reply:
Hi, I just though I would chime in here since I do use `whatshap stats` and think this would be a really nice addition to MultiQC. 

For me the  `phased / heterozygous_variants` ratio (i.e. % phased) would be most valuable for comparison. I also use`phased_snvs / heterozygous_snvs` to get the ratio for SNVs which is usually higher in my case. 

As for block size comparison I mostly rely on the `NG50` measure. 

For my part I don't mind having a lot of columns in the General Statistics table, I just have a config which selects the ones I am interested in. It would still nice to have a separate table for all `whatshap stats` statistics but maybe this could be a separate PR.

Reply:
Thanks for the feedback @pontushojer and @marcelm! I opted to include `% Phased Variants` and `Avg bp per Block` in the general statistics.  `NG50` is hidden by default, since it is not always set in the .tsv output file. I also added all statistics to a separate statistics table in the WhatsHap section of the report.